File: benchmark/binary_trees.js
/* The Great Computer Language Shootout
http://shootout.alioth.debian.org/
contributed by Isaac Gouy */

function TreeNode(left, right, item) {
  this.left = left;
  this.right = right;
  this.item = item;
}

TreeNode.prototype.itemCheck = function() {
  if (this.left == null) return this.item;
  return this.item + this.left.itemCheck() - this.right.itemCheck();
}

function bottomUpTree(item, depth) {
  if (depth > 0) {
    return new TreeNode(
        bottomUpTree(2 * item - 1, depth - 1),
        bottomUpTree(2 * item, depth - 1), item);
  }

  return new TreeNode(null, null, item);
}

var minDepth = 4;
var maxDepth = 14;
var stretchDepth = maxDepth + 1;

var start = process.hrtime();

var check = bottomUpTree(0, stretchDepth).itemCheck();
console.log("stretch tree of depth " + stretchDepth + "\t check: " + check);

var longLivedTree = bottomUpTree(0, maxDepth);
for (var depth = minDepth; depth <= maxDepth; depth += 2) {
  var iterations = 1 << (maxDepth - depth + minDepth);

  check = 0;
  for (var i = 1; i <= iterations; i++) {
    check += bottomUpTree(i, depth).itemCheck();
    check += bottomUpTree(-i, depth).itemCheck();
  }

  console.log(iterations * 2 + "\t trees of depth " + depth +
      "\t check: " + check);
}

console.log("long lived tree of depth " + maxDepth + "\t check: "
    + longLivedTree.itemCheck());

var elapsed = process.hrtime(start);
elapsed = elapsed[0] + elapsed[1] / 1000000000;
console.log("elapsed: " + elapsed);

File: benchmark/binary_trees.rb
# The Computer Language Shootout Benchmarks
# http://shootout.alioth.debian.org
#
# contributed by Jesse Millikan
# Modified by Wesley Moxam


def item_check(left, item, right)
  return item if left.nil?
  item + item_check(*left) - item_check(*right)
end

def bottom_up_tree(item, depth)
  return [nil, item, nil] unless depth > 0
  item_item = 2 * item
  depth -= 1
  [bottom_up_tree(item_item - 1, depth), item, bottom_up_tree(item_item, depth)]
end

max_depth = 12
min_depth = 4

max_depth = min_depth + 2 if min_depth + 2 > max_depth

stretch_depth = max_depth + 1
stretch_tree = bottom_up_tree(0, stretch_depth)

start = Time.now
puts "stretch tree of depth #{stretch_depth} check: #{item_check(*stretch_tree)}"
stretch_tree = nil

long_lived_tree = bottom_up_tree(0, max_depth)

min_depth.step(max_depth + 1, 2) do |depth|
  iterations = 2**(max_depth - depth + min_depth)

  check = 0

  for i in 1..iterations
    temp_tree = bottom_up_tree(i, depth)
    check += item_check(*temp_tree)

    temp_tree = bottom_up_tree(-i, depth)
    check += item_check(*temp_tree)
  end

  puts "#{iterations * 2} trees of depth #{depth} check: #{check}"
end

puts "long lived tree of depth #{max_depth} check: #{item_check(*long_lived_tree)}"
puts "elapsed: " + (Time.now - start).to_s
